app/spotify.py

```diff
@@ -22,34 +65,65 @@ def get_songs_from_spotify_website(playlist):
-        url = playlist.replace("spotify.com/playlist", "spotify.com/embed/playlist")
-    else:
-        url = base_url.format(playlist)
+    url_info = parse_uri(playlist)
 
-    req = requests.get(url)
+    req = requests.get(token_url)
     if req.status_code != 200:
-        raise SpotifyWebsiteParserException("ERROR: {} gave us not a 200. Instead: {}".format(url, req.status_code))
-
-    bs = BeautifulSoup(req.text, 'html.parser')
-    try:
-        songs_txt = bs.find('script', {'id': 'resource'}).string.strip()
-    except AttributeError:
-        raise SpotifyWebsiteParserException("ERROR: Could not get songs from Spotify website. Wrong Playlist id? Tried {}".format(url))
-    songs_json = json.loads(unquote(songs_txt))
-
-    for track in songs_json['tracks']['items']:
-        artist = track['track']['artists'][0]['name']
-        song = track['track']['name']
-        full = "{} {}".format(artist, song)
-        # remove everything in brackets to get better search results later on Deezer
-        # e.g. (Radio  Version) or (Remastered)
-        full = re.sub(r'\([^)]*\)', '', full)
-        return_data.append(full)
+        raise SpotifyWebsiteParserException(
+            "ERROR: {} gave us not a 200. Instead: {}".format(token_url, req.status_code))
+    token = req.json()
+
+    if url_info['type'] == "playlist":
+        url = playlist_base_url.format(url_info["id"])
+        while True:
+            resp = get_json_from_api(url, token["accessToken"])
+            for track in resp['items']:
+                return_data.append(parse_track(track["track"]))
+
+            if resp['next'] is None:
+                break
+            url = resp['next']
+    elif url_info["type"] == "track":
+        resp = get_json_from_api(track_base_url.format(url_info["id"]), token["accessToken"])
+        if resp is None:  # try again in case of rate limit
+            resp = get_json_from_api(track_base_url.format(url_info["id"]), token["accessToken"])
+
+        return_data.append(parse_track(resp))
+    elif url_info["type"] == "album":
+        resp = get_json_from_api(album_base_url.format(url_info["id"]), token["accessToken"])
+        if resp is None:  # try again in case of rate limit
+            resp = get_json_from_api(album_base_url.format(url_info["id"]), token["accessToken"])
+
+        for track in resp['items']:
+            return_data.append(parse_track(track))
+
     return return_data
 
 
+def parse_track(track):
+    artist = track['artists'][0]['name']
+    song = track['name']
+    full = "{} {}".format(artist, song)
+    # remove everything in brackets to get better search results later on Deezer
+    # e.g. (Radio  Version) or (Remastered)
+    return re.sub(r'\([^)]*\)', '', full)
+
+
+def get_json_from_api(api_url, access_token):
+    req = requests.get(api_url, headers={'Authorization': 'Bearer {}'.format(access_token)})
+
+    if req.status_code == 429:
+        sleep(int(req.headers.get("Retry-After")) + 1)
```

Comment:
I don't know how much rate limiting is a thing for the spotify api. If we got blocked regularly, maybe it's better to add a print statement here.

Reply:
During my development here I didn't get rate limited once. I haven't really been able to fully test it. Maybe a print statement is a good idea regardless. Will add.